### DISCUSSION TITLE
객체와 같은 참조타입은 왜 원시타입과 다르게 변경가능한 값으로 설계되었는지 설명해주세요.

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#45] [원시 값과 객체의 비교]
- 객체와 같은 참조타입은 왜 원시타입과 다르게 변경가능한 값으로 설계되었는지 설명해주세요. 객체가 불변성을 지키면 안되는걸까요?
  - 참조타입 값들은 어떻게 여러 타입이 가능한가요?
  - C언어의 경우, 배열에는 특정 타입의 데이터만 들어가는데요, 이러한 방식의 장점은 무엇인가요?

**Comment by pipisebastian** [comment]
> 객체와 같은 참조타입은 왜 원시타입과 다르게 변경가능한 값으로 설계되었는지 설명해주세요. 객체가 불변성을 지키면 안되는걸까요?

- 객체타입도 불변성을 지키려면, 많은 비용과 메모리가 듭니다. 원시타입의 경우 불변성을 지키기 위해, 아예 새로운 메모리 공간을 할당하는데, 크기가 큰 객체를 전부 복사하는 비용이 클 것입니다.

> 참조타입 값들은 어떻게 여러 타입이 가능한가요?
- 참조타입의 값, 예를 들어 배열에 들어가는 원소들은, 원시타입처럼 값을 직접 저장하는게 아니라, 값이 저장된 메모리 주소를 참조합니다. (참조타입의 배열이 메모리 주소를 참조하는 것 뿐만 아니라, 배열의 원소들도 메모리 주소를 참조. 한단계를 더 거치는 것)
- 즉, 배열 요소는 메모리 주소를 참조하므로, 서로 다른 타입의 데이터를 저장할 수 있습니다.

> C언어의 경우, 배열에는 특정 타입의 데이터만 들어가는데요, 이러한 방식의 장점은 무엇인가요? 
- 메모리 관리가 효율적입니다.
- 자바스크립트는, 참조타입에 무슨 값이 들어갈지 모르기 때문에, 여유로운 메모리 확보가 필요합니다. 반면 c언어는 모든 요소가 동일한 크기이기에, 미리 예측이 가능하여 효율적인 관리가 가능합니다.

**Comment by publdaze** [comment]
> - 객체와 같은 참조타입은 왜 원시타입과 다르게 변경가능한 값으로 설계되었는지 설명해주세요. 객체가 불변성을 지키면 안되는걸까요?
- 객체를 변경할 때 원시 값처럼 새로운 메모리 공간을 확보해 참조하는 식으로 처리한다면 명확하고 신뢰성이 확보되겠지만,
- 객체는 크기가 매우 클 수도 있고, 일정하지 않으며, 프로퍼티가 중첩 객체일 수도 있어서 새로운 메모리 공간을 다시 확보해 값을 저장하기 위해 복사를 할 때 많은 비용이 들 수 있음. 따라서 메모리를 효율적으로 사용하기 위해 변경 가능한 값으로 설계됨.
> - 참조타입 값들은 어떻게 여러 타입이 가능한가요?
- 참조 타입 값들은 변수에 직접 값이 저장되는 게 아닌 객체의 메모리 주소, 참조 값을 저장함. 실제 객체의 위치는 메모리 힙 내에 위치하며 메모리 힙에서는 런타임에 동적으로 할당이 가능하기 때문에 여러 타입을 사용하는 것이 가능함
> - C언어의 경우, 배열에는 특정 타입의 데이터만 들어가는데요, 이러한 방식의 장점은 무엇인가요?
- 메모리 크기가 고정이 되어 있기 때문에 예측이 쉬워지고, 정해진 메모리 공간을 사용할 수 있음
